Mathlib/CategoryTheory/Generator.lean

```diff
@@ -567,4 +569,98 @@ theorem wellPowered_of_isDetector [HasPullbacks C] (G : C) (hG : IsDetector G) :
   haveI := small_subsingleton ({G} : Set C)
   wellPowered_of_isDetecting hG
 
+theorem wellPowered_of_isSeparator [HasPullbacks C] [Balanced C] (G : C) (hG : IsSeparator G) :
+    WellPowered C := wellPowered_of_isDetecting hG.isDetector
+
+section HasGenerator
+
+variable (E : Type u₃) [Category.{v₃} E]
+
+/--
+A category instantiates `HasSeparator` if and only if it has a separator.
+See also `IsSeparator`.
+-/
+class HasSeparator : Prop :=
+  hasSeparator : ∃ G : E, IsSeparator G
+
+/--
+A category instantiates `HasCoseparator` if and only if it has a coseparator.
+See also `IsCoseparator`.
+-/
+class HasCoseparator : Prop :=
+  hasCoseparator : ∃ G : E, IsCoseparator G
+
+/--
+A category instantiates `HasDetector` if and only if it has a detector.
+See also `IsDetector`.
+-/
+class HasDetector : Prop :=
+  hasDetector : ∃ G : E, IsDetector G
+
+/--
+A category instantiates `HasCodetector` if and only if it has a codetector.
+See also `IsCodetector`.
+-/
+class HasCodetector : Prop :=
+  hasCodetector : ∃ G : E, IsCodetector G
+
+theorem HasSeparator.hasDetector [Balanced C] [HasSeparator C] : HasDetector C := by
+  obtain ⟨G, hG⟩ : HasSeparator C := inferInstance
+  exact ⟨G, hG.isDetector⟩
+
+theorem HasDetector.hasSeparator [HasEqualizers C] [HasDetector C] : HasSeparator C := by
+  obtain ⟨G, hG⟩ : HasDetector C := inferInstance
+  exact ⟨G, hG.isSeparator⟩
+
+theorem HasCoseparator.hasCodetector [Balanced C] [HasCoseparator C] : HasCodetector C := by
+  obtain ⟨G, hG⟩ : HasCoseparator C := inferInstance
+  exact ⟨G, hG.isCodetector⟩
+
+theorem HasCodetector.hasCoseparator [HasCoequalizers C] [HasCodetector C] : HasCoseparator C := by
+  obtain ⟨G, hG⟩ : HasCodetector C := inferInstance
+  exact ⟨G, hG.isCoseparator⟩
+
+theorem HasDetector.wellPowered [HasPullbacks C] [HasDetector C] : WellPowered C := by
+  obtain ⟨G, hG⟩ : HasDetector C := inferInstance
+  exact wellPowered_of_isDetector G hG
+
+theorem HasSeparator.wellPowered [HasPullbacks C] [Balanced C] [HasSeparator C] :
+    WellPowered C := HasSeparator.hasDetector.wellPowered
+
+section Dual
+
+@[simp]
+theorem hasSeparator_op_iff : HasSeparator Eᵒᵖ ↔ HasCoseparator E :=
+  ⟨fun ⟨G, hG⟩ => ⟨unop G, (isCoseparator_unop_iff G).mpr hG⟩,
+   fun ⟨G, hG⟩ => ⟨op G, (isSeparator_op_iff G).mpr hG⟩⟩
+
+@[simp]
+theorem hasCoseparator_op_iff : HasCoseparator Eᵒᵖ ↔ HasSeparator E :=
+  ⟨fun ⟨G, hG⟩ => ⟨unop G, (isSeparator_unop_iff G).mpr hG⟩,
+   fun ⟨G, hG⟩ => ⟨op G, (isCoseparator_op_iff G).mpr hG⟩⟩
+
+theorem HasSeparator.hasCoseparator_op [HasSeparator C] : HasCoseparator Cᵒᵖ := by simp [*]
```

Comment:
`hasCoseparator_op` and `hasSeparator_op` should be instances (and similarly for detectors below).